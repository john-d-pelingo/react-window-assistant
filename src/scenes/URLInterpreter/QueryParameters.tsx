import { Table, TableBody, TableCell, TableRow } from '@material-ui/core'
import { ParsedQuery } from 'query-string'
import React, { FC } from 'react'

interface IQueryParametersProps {
  parsedQuery: ParsedQuery
}

export const QueryParameters: FC<IQueryParametersProps> = ({ parsedQuery }) => (
  <Table>
    <TableBody>
      <TableRow>
        <TableCell>Key</TableCell>
        <TableCell>Value</TableCell>
      </TableRow>
      {Object.entries(parsedQuery).map(([key, value], ii) => (
        <TableRow key={`${key}-${ii}`}>
          <TableCell>{key}</TableCell>
          <TableCell>
            {Array.isArray(value) ? value.join(',') : value}
          </TableCell>
        </TableRow>
      ))}
    </TableBody>
  </Table>
)
