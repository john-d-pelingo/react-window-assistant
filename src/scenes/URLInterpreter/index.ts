export { URLInterpreter as default } from './URLInterpreter'
