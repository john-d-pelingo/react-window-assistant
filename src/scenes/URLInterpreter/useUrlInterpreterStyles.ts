import { createStyles, makeStyles } from '@material-ui/core'

export const useUrlInterpreterStyles = makeStyles(theme =>
  createStyles({
    container: {
      margin: `${theme.spacing(12.5)}px 0`,
    },
  }),
)
