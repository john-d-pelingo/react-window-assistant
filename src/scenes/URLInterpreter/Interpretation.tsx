import {
  Paper,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
} from '@material-ui/core'
import { parse } from 'query-string'
import React, { FC } from 'react'
import styled from 'styled-components'

import { QueryParameters } from './QueryParameters'

const StyledPaper = styled(Paper)`
  width: 100%;
  margin-top: 24px;
  overflow-x: auto;
`

interface IInterpretationProps {
  urlInstance: URL
}

export const Interpretation: FC<IInterpretationProps> = ({ urlInstance }) => {
  const parsedQuery = parse(urlInstance.search)

  return (
    <StyledPaper>
      <Table style={{ minWidth: 650 }}>
        <TableHead>
          <TableRow>
            <TableCell colSpan={2}>Interpretation</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          <TableRow>
            <TableCell>Origin</TableCell>
            <TableCell>{urlInstance.origin}</TableCell>
          </TableRow>
          <TableRow>
            <TableCell>Pathname</TableCell>
            <TableCell>{urlInstance.pathname}</TableCell>
          </TableRow>
          {Boolean(urlInstance.hash) && (
            <TableRow>
              <TableCell>Hash</TableCell>
              <TableCell>{urlInstance.hash}</TableCell>
            </TableRow>
          )}
          {Boolean(urlInstance.port) && (
            <TableRow>
              <TableCell>Port</TableCell>
              <TableCell>{urlInstance.port}</TableCell>
            </TableRow>
          )}
          {Boolean(urlInstance.search) && (
            <TableRow>
              <TableCell>Search</TableCell>
              <TableCell>{urlInstance.search}</TableCell>
            </TableRow>
          )}
          {Object.keys(parsedQuery).length > 0 && (
            <TableRow>
              <TableCell>Search Parameters</TableCell>
              <TableCell>
                <QueryParameters parsedQuery={parsedQuery} />
              </TableCell>
            </TableRow>
          )}
        </TableBody>
      </Table>
    </StyledPaper>
  )
}
