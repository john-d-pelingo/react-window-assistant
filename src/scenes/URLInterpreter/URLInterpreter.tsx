import { Container, TextField } from '@material-ui/core'
import { RouteComponentProps } from '@reach/router'
import React, { ChangeEvent, FC, useEffect, useRef, useState } from 'react'

import { urlInterpreterQueryParameter } from 'consants/strings'
import { appendQueryParameter } from 'helpers/appendQueryParameter'
import { extractQueryParameter } from 'helpers/extractQueryParameter'

import { Interpretation } from './Interpretation'
import { useUrlInterpreterStyles } from './useUrlInterpreterStyles'

export const URLInterpreter: FC<RouteComponentProps> = () => {
  const classes = useUrlInterpreterStyles()
  const inputNode = useRef<HTMLInputElement>()
  const [urlText, setUrlText] = useState('')
  const [urlInstance, setUrlInstance] = useState<URL | null>()

  useEffect(() => {
    if (inputNode.current) {
      inputNode.current.focus()
    }

    const queryParameter = extractQueryParameter(urlInterpreterQueryParameter)

    if (queryParameter) {
      setUrlText(queryParameter)

      try {
        const newUrlInstance = new URL(queryParameter)
        setUrlInstance(newUrlInstance)
      } catch (error) {
        return
      }
    }
  }, [])

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>): void => {
    const { value } = event.target
    setUrlText(value)

    try {
      const newUrlInstance = new URL(value)
      setUrlInstance(newUrlInstance)
      appendQueryParameter(newUrlInstance.href)
    } catch (error) {
      appendQueryParameter(value)
      setUrlInstance(null)
    }
  }

  return (
    <Container className={classes.container} maxWidth="xl">
      <TextField
        fullWidth
        inputProps={{
          'aria-label': 'URL input',
        }}
        inputRef={inputNode}
        margin="normal"
        onChange={handleInputChange}
        placeholder="URI to interpret"
        value={urlText}
      />
      {urlInstance && <Interpretation urlInstance={urlInstance} />}
    </Container>
  )
}
