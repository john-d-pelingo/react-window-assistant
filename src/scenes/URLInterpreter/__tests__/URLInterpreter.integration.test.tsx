import 'jest-dom/extend-expect'

import { globalHistory } from '@reach/router'
import { fireEvent, render } from '@testing-library/react'
import React from 'react'

import { URLInterpreter } from '../URLInterpreter'

// TODO: refactor to use my own history e.g. https://reach.tech/router/api/createHistory
// instead of `globalHistory`. Use the created `history` object with `LocationProvider`.
// And then when the `history` object is needed, import it from the file that creates
// the new `history` object. Pass that object to functions that require it.

describe.skip('scenes - URLInterpreter', () => {
  it('displays info about a valid URL', () => {
    expect(globalHistory.location.href).toBe('http://localhost/')

    const { getByLabelText, getByText, queryByText } = render(
      <URLInterpreter />,
    )

    const urlInputElement = getByLabelText('URL input') as HTMLInputElement

    expect(document.activeElement).toEqual(urlInputElement)

    fireEvent.change(urlInputElement, {
      target: {
        value:
          'https://facebook.github.io:420/create-react-app/docs/using-the-public-folder?query=1&query=65&gg=0#docsNav',
      },
    })

    expect(urlInputElement.value).toBe(
      'https://facebook.github.io:420/create-react-app/docs/using-the-public-folder?query=1&query=65&gg=0#docsNav',
    )
    expect(queryByText(/interpretation/i)).toBeInTheDocument()
    expect(getByText('Origin').nextSibling!.textContent).toBe(
      'https://facebook.github.io:420',
    )
    expect(getByText('Pathname').nextSibling!.textContent).toBe(
      '/create-react-app/docs/using-the-public-folder',
    )
    expect(getByText('Hash').nextSibling!.textContent).toBe('#docsNav')
    expect(getByText('Port').nextSibling!.textContent).toBe('420')
    expect(getByText('Search').nextSibling!.textContent).toBe(
      '?query=1&query=65&gg=0',
    )
    expect(queryByText(/search parameters/i)).toBeInTheDocument()
    expect(getByText('query').nextSibling!.textContent).toBe('1,65')

    expect(globalHistory.location.search).toContain(
      '?url=https%3A%2F%2Ffacebook.github.io%3A420%2Fcreate-react-app%2Fdocs%2Fusing-the-public-folder%3Fquery%3D1%26query%3D65%26gg%3D0%23docsNav',
    )
  })

  it("doesn't display info about an invalid URL", () => {
    expect(globalHistory.location.href).toBe('http://localhost/')

    const { getByLabelText, queryByText } = render(<URLInterpreter />)

    const urlInputElement = getByLabelText('URL input') as HTMLInputElement

    fireEvent.change(urlInputElement, {
      target: {
        value: "What's)this!U+R>L%(_[≥æ™‘«¢“∞æ™æ¡",
      },
    })

    expect(urlInputElement.value).toBe("What's)this!U+R>L%(_[≥æ™‘«¢“∞æ™æ¡")

    expect(queryByText(/interpretation/i)).not.toBeInTheDocument()
    expect(window.location.href).toContain(
      '?query=What%27s%29this%21U%2BR%3EL%25%28_%5B%E2%89%A5%C3%A6%E2%84%A2%E2%80%98%C2%AB%C2%A2%E2%80%9C%E2%88%9E%C3%A6%E2%84%A2%C3%A6%C2%A1',
    )
  })
})
