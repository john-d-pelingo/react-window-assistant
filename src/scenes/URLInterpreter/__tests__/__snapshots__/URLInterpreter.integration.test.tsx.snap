// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scenes - URLInterpreter mounts 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthXl"
  style="margin: 100px 0px;"
>
  <div
    class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        aria-label="URL input"
        class="MuiInputBase-input MuiInput-input"
        placeholder="URI to interpret"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
